R/extract_param.R

```diff
@@ -42,103 +47,248 @@ extract_param <- function(type = c("percentiles", "range"),
+#' median and range of a sample and the number of samples
+#'
+#' @inheritParams extract_param
+#'
+#' @return A list with output from stats::optim. See ?optim for more details
+#' @export
+#'
+#' @examples
+#' # extract parameters of a gamma distribution from median and range
+#' extract_param_range(
+#'   values = c(8, 6, 13),
+#'   distribution = "gamma",
+#'   samples = 20
+#' )
+extract_param_range <- function(values,
+                                distribution,
+                                samples) {
+
+  # check input
+  distribution <- match.arg(
+    arg = distribution,
+    choices = c("lnorm", "gamma", "weibull"),
+    several.ok = FALSE
+  )
+  checkmate::assert_numeric(values)
+  checkmate::assert_numeric(samples, lower = 1)
+
+  # Set initial values for optimisation
+  param <- stats::runif(n = 2, min = 0, max = 5)
+  names(values) <- c("median", "lower", "upper")
+  values_in <- c(values, n = samples)
+
+  if (distribution == "lnorm") {
+    names(param) <- c("meanlog", "sdlog")
+    optim_params <- stats::optim(
+      param,
+      fit_function_lnorm_range,
+      method = "L-BFGS-B",
+      val = values_in,
+      lower = c(-1e5, 1e-10)
+    )
+  }
+  if (distribution == "gamma") {
+    names(param) <- c("shape", "scale")
+    optim_params <- stats::optim(
+      param,
+      fit_function_gamma_range,
+      method = "L-BFGS-B",
+      val = values_in,
+      lower = c(1e-10, 1e-10)
+    )
   }
+  if (distribution == "weibull") {
+    names(param) <- c("shape", "scale")
+    optim_params <- stats::optim(
+      param,
+      fit_function_weibull_range,
+      method = "L-BFGS-B",
+      val = values_in,
+      lower = c(1e-10, 1e-10)
+    )
+  }
+  optim_params
+}
 
-  # Output parameters
-  result2$par
+#' Checks whether the optimisation of distribution parameters has converged to
+#' stable value for the parameters and function output for multiple iterations
+#'
+#' @description This function is a try and prevent optimisation to a local
+#' optim and thus checks whether multiple optimisation routines are consistently
+#' finding parameter values to within a set tolerance of 1e-5.
+#'
+#' @param optim_params_list A list, where each element is the output of
+#' stats::optim. See ?optim for more details
+#' @param optim_params A list given by the output of stats::optim
+#' @param optim_conv A boolean value indicating whether the optimisation has
+#' converged over multiple iterations
+#'
+#' @return Boolean
+check_optim_conv <- function(optim_params_list,
+                             optim_params,
+                             optim_conv) {
+
+  # no pairwise comparison on first iterations
+  if (length(optim_params_list) > 1) {
+
+    # extract parameters from list
+    params <- lapply(optim_params_list, "[[", 1)
+
+    # calculate pairwise comparison across all iterations
+    param_a_dist <- stats::dist(unlist(lapply(params, "[[", 1)))
+    param_b_dist <- stats::dist(unlist(lapply(params, "[[", 2)))
```

Comment:
I used pairwise distances over sequential distances because `check_optim_conv()` looks to see whether a global optimum has been attained. I initially coded this sequentially, but if for example the global optimum is found on the first run and third run it wouldn't stop the loop; and would wait until two successive answers are likely the optimum. Therefore, the additional cost of running `dist()` should be recouped by running less iterations. Another reason why it isn't comparing consecutive values is because each iteration starts from randomly chosen initial parameters, and not from the output of the last round of optimisation (as might be done by other algorithms).

Please let me know if this makes sense and whether you agree with the logic. There is nothing exact about this function, it is just to help prevent times when the optimisation might return a local optimum on its first run.

I prefer the use of `unlist(lapply())` over `sapply()`, as I've had cases in the past where `sapply()` didn't return a vector (I don't think I looked into why). Also a small speed bonus: https://stackoverflow.com/questions/18761541/why-is-unlistlapply-faster-than-sapply

Reply:
Yes all of that makes sense. I didn't catch that it wasn't starting each iteration from the previous iteration's estimate. I'm curious about how its performance relates to the arguments of `extract_param`, but that's besides the point for this PR.

Fair enough with `unlist(lapply(...))`; `vapply` also exists as a vector return alternative, but may also be slower.